Add vehicle rental overlay

evansiroky opened the pull request:
This adds the ability to display rental vehicles as an overlay on the map.

landonreed commented on packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.js:
```diff
@@ -0,0 +1,259 @@
+  .map(makeRandomCarRentalStation);
+const randomEScooterStations = Array(20)
+  .fill(null)
+  .map(makeRandomEScooterRentalStation);
```
I'd like to see a lot of this mock data perhaps live in a `__mocks__` folder for this package. Isn't that the pattern we've been generally following?

fpurcell replied:
Honestly, I would rather we get away from generating random data, and used canned data for the Storybooks (unless there's a storybook that absolutely needs a current datestamp on the mock data), @evansiroky.

Three reasons:
  a) when folks see a stop or vehicle in the Willamette (on storybook), they wonder if the app is broken (as @c-harper can attest :-).
  b) the format of the mock'd data might not fully reflect a snapshot of the real data the component might be fetching in the future  ... so having a static __mocks__/blah.json file pulled from the service, and rendering that seems to get one closer to the eventual truth
  c) I've heard that some recent changes to Storybook makes for a good unit testing framework (via a plugin or something -- haven't found out the how yet), and was thinking that if that does exist, it would probably need static data.

landonreed commented on packages/vehicle-rental-overlay/src/styled.js:
```diff
@@ -0,0 +1,62 @@
+  height: 24px;
+`;
+
+export const hubIcons = [
```
Maybe this could use a comment describing what the icon is.

landonreed commented on packages/vehicle-rental-overlay/src/index.js:
```diff
@@ -0,0 +1,251 @@
+  })
+);
+
+class VehicleRentalOverlay extends MapLayer {
```
Definitely deserves JSdoc given that it serves as the overlay for a number of different modes.

landonreed commented on packages/vehicle-rental-overlay/src/index.js:
```diff
@@ -0,0 +1,251 @@
+  updateLeafletElement() {}
+
+  startRefreshing() {
+    // ititial station retrieval
```
typo: ititial

landonreed commented on packages/vehicle-rental-overlay/src/index.js:
```diff
@@ -0,0 +1,251 @@
+      stationName = `${stationNetworks} ${stationName}`;
+    } else if (station.isFloatingVehicle) {
+      // assumes that all floating vehicles are E-scooters
+      stationName = `${stationNetworks} E-scooter`;
```
Since we're touching this code, it might be worth extracting the `E-scooter` string into a prop that can be overridden if need be.

landonreed commented on packages/vehicle-rental-overlay/src/index.js:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    if (!filteredStations || filteredStations.length === 0)
+      return <FeatureGroup />;
```
Does the airbnb style guide really recommend not encapsulating multi-line conditionals in curly braces? Not sure I like that at all.

binh-dam-ibigroup replied:
On this style rule, it seems more like when you write a long 
`if (condition) thenStatement();` line, then es-lint recommends to put the thenStatement() into its own line. In that case would also add braces around thenStatement() to avoid confusion.

fpurcell commented on packages/core-utils/src/types.js:
```diff
@@ -20,23 +27,65 @@ export const languageConfigType = PropTypes.shape({
- * Represents the expected configuration of the webapp
+ * Represents the expected configuration of the webapp.
+ *
+ * Note: this is an incomplete type mapping.
```
What all is missing? ... maybe list here or the line comment below.

evansiroky replied:
There are probably too many things missing and the config itself is sometimes a moving target that may not be worth the time to make a detailed inventory. Ideally this type shouldn't even be used directly in these components that we're extracting. Only parts of the config should be used in each component.

fpurcell commented on packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.js:
```diff
@@ -0,0 +1,259 @@
+
+function makeRandomCarRentalStation(mapVal, idx) {
+  // make a free-floating style car rental station
+  const id = `Bike ${idx}`;
```
Minor: storybook popup says car2go BIKE :-)

binh-dam-ibigroup commented on packages/core-utils/src/types.js:
```diff
@@ -20,23 +27,65 @@ export const languageConfigType = PropTypes.shape({
   stopViewer: PropTypes.string
 });
 
+export const mapSymbolsType = PropTypes.arrayOf(
```
This name seems too generic and IMO should instead suggest why the dockStrokeColor is a relevant attribute.

binh-dam-ibigroup commented on packages/core-utils/src/types.js:
```diff
@@ -243,6 +306,25 @@ export const transitIndexStopWithRoutes = PropTypes.shape({
+ * This models data about vehicle rental stations as obtained from various
+ * vehicle rental API endpoints from OTP.
+ */
+export const stationsType = PropTypes.arrayOf(
```
Did you mean `stationType`? (It seems to be about one station as opposed to multiple.)

evansiroky replied:
No, it's specifically for multiple as indicated by the arrayOf propType.

binh-dam-ibigroup replied:
Oh I missed the inner array. Then do you want to do a type for one station, and have the stations props to be PropTypes.arrayOf(stationType)?

binh-dam-ibigroup commented on packages/from-to-location-picker/src/index.js:
```diff
@@ -1,59 +1,100 @@
+      onFromClick();
+      return;
+    }
+    setLocation({
```
Does the code that handles setLocation split into two different paths for 'from' and 'to'?

evansiroky replied:
This is somewhat of an artifact of extracting stuff from otp-react-redux. Here is the relevant action: https://github.com/opentripplanner/otp-react-redux/blob/906d9226553cbbdd83cf455986f199c5a6964f72/lib/actions/map.js#L38-L63

evansiroky commented:
All right, I just did some refactoring to address all of these comments as follows:

- Used a mocked static list of vehicle rentals
- Added more comments
- Fixed a typo
- Made a slot for a custom vehicle station name function
- Renamed the mapSymbols type to a more meaningful name
- Extracted each styled hub icon styled component to allow for further individualized customization
- Made a story demonstrating use of the custom vehicle station name function

evansiroky commented:
@binh-dam-ibigroup, I just updated the station type as you requested.

binh-dam-ibigroup commented:
Your call if you want to wait for Landon or not.
